### PR TITLE
  Add more transaction metrics 

### PR DESCRIPTION
**Pull request by davidegrohmann**

### CODE REVIEW

**Comment by davidegrohmann** [comment]
Needs to be based on top of https://github.com/neo4j/neo4j/pull/5888

**Comment by davidegrohmann** [comment]
@tinwelint green again!

File: community/kernel/src/test/java/org/neo4j/kernel/impl/transaction/TransactionCountersChecker.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.transaction;
+
+import static org.junit.Assert.assertEquals;
+
+class TransactionCountersChecker
```
Comment (MishaDemianenko) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/transaction/TransactionCountersChecker.java:24]: TransactionCountersTest?

Comment (MishaDemianenko) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/transaction/TransactionCountersChecker.java:24]: ah no found...just ignore

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java [hunk 2]
```diff
@@ -261,7 +249,10 @@ public void markForTermination()
         {
             failure = true;
             terminated = true;
-            transactionMonitor.transactionTerminated();
+            if ( !closed )
+            {
+                transactionMonitor.transactionTerminated( txState != null );
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java]: txState state check should also include hasChanges() check

File: community/kernel/src/main/java/org/neo4j/kernel/impl/factory/PlatformModule.java [hunk 3]
```diff
@@ -285,9 +285,9 @@ protected PageCache createPageCache( FileSystemAbstraction fileSystem,
         return pageCache;
     }
 
-    protected TransactionCounters createTransactionCounters()
+    protected TransactionStats createTransactionCounters()
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/factory/PlatformModule.java]: please align method name and structure name

File: community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/TransactionCounters.java [hunk 4]
```diff
@@ -19,72 +19,33 @@
  */
 package org.neo4j.kernel.impl.transaction;
 
-import java.util.concurrent.atomic.AtomicLong;
-
-public class TransactionCounters implements TransactionMonitor
+public interface TransactionCounters
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/TransactionCounters.java:22]: do we really need to introduce this interface? what will be the other possible implementation?

File: enterprise/metrics/src/main/java/org/neo4j/metrics/source/CheckPointingMetrics.java [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    {
+        // check pointing
+        if ( config.get( MetricsSettings.neoCheckPointingEnabled ) )
+        {
+            registry.remove( CHECK_POINT_EVENTS );
+            registry.remove( CHECK_POINT_TOTAL_TIME );
+        }
+
+        // log rotation
+        if ( config.get( MetricsSettings.neoCheckPointingEnabled ) )
```
Comment (MishaDemianenko) [inline enterprise/metrics/src/main/java/org/neo4j/metrics/source/CheckPointingMetrics.java]: wrong condition, cpd :)

**Comment by MishaDemianenko** [comment]
@davidegrohmann couple of comments, please avoid code reformatting that is not related to a commit, it's really hard to read.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/TransactionCounters.java (hunk 4 above)
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/TransactionCounters.java:22]: Yes I think we do need it. Indeed having one interface for reading counters allows to us not to leak the `TransactionMonitor` methods where they are not supposed to be used.

File: enterprise/metrics/src/main/java/org/neo4j/metrics/source/CheckPointingMetrics.java (hunk 5 above)
Comment (davidegrohmann) [inline enterprise/metrics/src/main/java/org/neo4j/metrics/source/CheckPointingMetrics.java]: :scream:

File: community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/TransactionStats.java [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    @Override
+    public void transactionStarted()
+    {
+        startedTransactionCount.incrementAndGet();
+        long active = activeReadTransactionCount.incrementAndGet();
+        peakTransactionCount = Math.max( peakTransactionCount, active );
+    }
+
+    @Override
+    public void transactionFinished( boolean successful, boolean write )
```
Comment (lutovich) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/TransactionStats.java]: Probably better to rename `successful` to `committed`.

File: enterprise/metrics/src/main/java/org/neo4j/metrics/source/CheckPointingMetrics.java [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+        this.registry = registry;
+        this.config = config;
+        this.checkPointerMonitor = checkPointerMonitor;
+        this.logRotationMonitor = logRotationMonitor;
+    }
+
+    @Override
+    public void start() throws Throwable
+    {
+        // check pointing
```
Comment (lutovich) [inline enterprise/metrics/src/main/java/org/neo4j/metrics/source/CheckPointingMetrics.java:62]: Such comments usually mean that 'extract method' should be done.

**Comment by tinwelint** [comment]
@davidegrohmann needs rebase and there are test failures.

File: enterprise/metrics/src/main/java/org/neo4j/metrics/source/CheckPointingMetrics.java (hunk 7 above)
Comment (davidegrohmann) [inline enterprise/metrics/src/main/java/org/neo4j/metrics/source/CheckPointingMetrics.java:62]: This has been fixed in another PR.

**Comment by davidegrohmann** [comment]
test failures should be fixed when rebasing on top of https://github.com/neo4j/neo4j/pull/5945

**Comment by davidegrohmann** [comment]
@tinwelint green again!

**Pull request merged**